### PR TITLE
k8s: fixed compilation of tests

### PR DESCRIPTION
**Pull request by hunchback**
broken due to changes done in topology/graph/graph.go

maybe we should consider re-adding k8s tests to list of defaults

### CODE REVIEW

**Comment by hunchback** [comment]
run skydive-k8s-tests

File: topology/probes/k8s/probe.go [hunk 1]
```diff
@@ -94,6 +94,11 @@ func (p *Probe) AppendNamespaceLinkers(types ...string) {
 
 // NewProbe creates the probe for tracking k8s events
 func NewProbe(g *graph.Graph, manager string, subprobes map[string]Subprobe, linkers []probe.Probe) *Probe {
+	names := []string{}
+	for k, _ := range subprobes {
+		names = append(names, k)
+	}
+	logging.GetLogger().Infof("probe %s subprobes %v", manager, names)
```
Comment (lebauce) [inline topology/probes/k8s/probe.go]: s/probe/Probe/

Comment (hunchback) [reply topology/probes/k8s/probe.go]: ok

**Comment by hunchback** [comment]
run skydive-k8s-tests

**Comment by hunchback** [comment]
run skydive-k8s-tests

**Comment by hunchback** [comment]
run skydive-k8s-tests

**Comment by hunchback** [comment]
@lebauce - can you assist with k8s failure:

```
panic: test timed out after 20m0s
```

File: topology/graph/indexer.go [hunk 2]
```diff
@@ -204,7 +204,7 @@ func NewMetadataIndexer(g *Graph, listenerHandler ListenerHandler, m ElementMatc
 		Indexer: NewIndexer(g, listenerHandler, func(n *Node) (kv map[string]interface{}) {
 			if match := n.MatchMetadata(m); match {
 				kv = make(map[string]interface{})
-				if values, err := getFieldsAsArray(n, indexes); err == nil && len(values) > 0 {
+				if values, err := getFieldsAsArray(n, indexes); err == nil {
```
Comment (lebauce) [inline topology/graph/indexer.go]: Why ???

**Comment by lebauce** [comment]
Why do we have to change `resourceHandlers` and `EnabledSubprobes` again ?

**Comment by lebauce** [comment]
@hunchback The test that timeouts is `TestK8sClusterNode` so the problem is probably related to the K8s probe (or this PR). I don't think this is a CI issue

**Comment by hunchback** [comment]
so i can reuse by istio

**Comment by hunchback** [comment]
the commit broke links between cluster to its children

**Comment by hunchback** [comment]
will check ...

File: topology/probes/istio/istio.go [hunk 3]
```diff
@@ -49,15 +52,22 @@ func NewIstioProbe(g *graph.Graph) (*k8s.Probe, error) {
...
+		"gateway":          newGatewayProbe,
+		"quotaspec":        newQuotaSpecProbe,
+		"quotaspecbinding": newQuotaSpecBindingProbe,
+		"serviceentry":     newServiceEntryProbe,
+		"virtualservice":   newVirtualServiceProbe,
+	}
+
+	subprobes := make(map[string]k8s.Subprobe)
+	for name, handler := range resourceHandlers {
+		subprobes[name] = handler(client, g)
```
Comment (lebauce) [inline topology/probes/istio/istio.go]: All the subprobes are created even if they are not enabled

**Comment by hunchback** [comment]
run skydive-k8s-tests

**Comment by hunchback** [comment]
@lebauce  - I think we may need to increase k8s execution time - later we can optimize and make the test run faster

File: topology/probes/istio/istio.go (hunk 3 above)
Comment (hunchback) [reply topology/probes/istio/istio.go]: ok - have corrected

**Comment by hunchback** [comment]
> @hunchback The test that timeouts is `TestK8sClusterNode` so the problem is probably related to the K8s probe (or this PR). I don't think this is a CI issue

this test PASSes when run locally on my devmode VM

**Comment by hunchback** [comment]
run skydive-k8s-tests

**Comment by hunchback** [comment]
run skydive-functional-tests-backend-elasticsearch

**Comment by lebauce** [comment]
run skydive-k8s-tests

File: topology/graph/indexer.go (hunk 2 above)
Comment (lebauce) [reply topology/graph/indexer.go]: @hunchback That's why before 0b81a07 we had a special case when len(indexes) == 0. Anyway, I'm against removing && len(values) > 0 as it simply makes no sense to have an empty list (values) in kv.

**Comment by hunchback** [comment]
but this breaks k8s which previously worked ...

**Comment by hunchback** [comment]
so are you suggesting i replace the indexlinker and implement my custor
aablinks  balinks methods?

**Comment by lebauce** [comment]
@hunchback Yes, but we were indexing all the objets that matched the filters,**even if they did not have the specified indices** . So basically all the nodes, with an empty values in the hashmap. Restoring the `case 0` you removed should fix Kubernetes

**Comment by hunchback** [comment]
not sure i understand ... what is the fix you propose needed to get cluster
links working again ?

**Comment by hunchback** [comment]
but this breaks cluster as indexer.Get() will not return nothing-nil
(rather than everything)

**Comment by lebauce** [comment]
@hunchback Then yes. Creating a metadata indexer for a single node (cluster) makes little sense. Every time an object that needs to be linked to the cluster appears, we were querying the indexer that had a "nil" hash. That returned all the nodes : the cluster node in our case.

I think for the cluster linker, a classic linker that implements the GetBALinks() is simpler : we just return a link from the cluster node (that we can simply store as an attribute of the clusterLinker) to the node.

**Comment by hunchback** [comment]
ok, will do

**Comment by hunchback** [comment]
@lebauce - I want to propose the following:

(1) k8s is currently broken and the k8s tests disabled blocking k8s/istio PRs - I would be happy if this currently PR could be merged as-is.

(2) regarding your proposition - will your change not also break listing on the children nodes (namespaces, etc.)?

```
func (linker *clusterLinker) GetABLinks(clusterNode *graph.Node) (edges []*grpah.Edge) {
        nodes := linker.ObjectIndexer.Get()           <===== wi
        for _, n := nodes {
                // append to edges
        }
        return
}
```

**Comment by lebauce** [comment]
@hunchback No need for GetABLinks as the cluster node already exists when the linker is started. You simply need to retrieve it and keep it as an attribute

**Comment by hunchback** [comment]
run skydive-functional-tests-backend-elasticsearch

**Comment by hunchback** [comment]
run skydive-k8s-tests

File: topology/probes/k8s/cluster.go [hunk 4]
```diff
@@ -27,47 +27,79 @@ import (
...
 }
 
-func newClusterProbe(clientset *kubernetes.Clientset, g *graph.Graph) Subprobe {
-	return &clusterCache{
+func newClusterProbe(clientset interface{}, g *graph.Graph) Subprobe {
+	c := &clusterCache{
 		EventHandler: clusterEventHandler,
 		graph:        g,
 	}
+	c.addClusterNode()
```
Comment (lebauce) [inline topology/probes/k8s/cluster.go]: This can be moved to the Start method

File: topology/probes/k8s/cluster.go [hunk 5]
```diff
@@ -27,47 +27,79 @@ import (
...
+
+// GetBALinks returns all the incoming links for a node
+func (linker *clusterLinker) GetBALinks(objectNode *graph.Node) (edges []*graph.Edge) {
+	edges = append(edges, linker.createEdge(clusterNode, objectNode))
+	return
 }
 
 func newClusterLinker(g *graph.Graph, manager string, types ...string) probe.Probe {
-	clusterIndexer := graph.NewMetadataIndexer(g, clusterEventHandler, graph.Metadata{"Manager": Manager, "Type": "cluster"})
+	clusterIndexer := newObjectIndexerFromFilter(g, g, newTypesFilter(Manager, "cluster"))
```
Comment (lebauce) [inline topology/probes/k8s/cluster.go]: This doesn't seem to be used any more. This can be removed an nil be passed to NewResourceLinker instead

File: topology/probes/k8s/cluster.go [hunk 6]
```diff
@@ -27,47 +27,79 @@ import (
...
+	objectIndexer *graph.MetadataIndexer
+}
+
+func (linker *clusterLinker) createEdge(cluster, object *graph.Node) *graph.Edge {
+	id := graph.GenID(string(cluster.ID), string(object.ID))
+	return linker.g.CreateEdge(id, cluster, object, topology.OwnershipMetadata(), graph.TimeUTC(), "")
+}
+
+// GetABLinks returns all the incoming links for a node
+func (linker *clusterLinker) GetABLinks(objectNode *graph.Node) (edges []*graph.Edge) {
```
Comment (lebauce) [inline topology/probes/k8s/cluster.go]: You can embed graph.DefaultListener in clusterLinker and remove the GetABLinks

File: topology/probes/k8s/cluster.go (hunk 4 above)
Comment (hunchback) [reply topology/probes/k8s/cluster.go]: if moved to the start then we may miss handling of node Add events:
- as k8s.Probe Start() first starts the links
- then k8s.Probe.Start() Starts the subprobes in *random* order (looping via a go map works that way).

Thus if for example `namespace` will start before `cluster` then clusterLinker.GetBALinks() would get a `namespace` creation event *before* the singleton cluster node exists and thus would not be able to add the edge. As we are not implementing `clusterLinker.GetABLinks()` we have no chance to add the missing edge ....

File: topology/probes/k8s/cluster.go (hunk 5 above)
Comment (hunchback) [reply topology/probes/k8s/cluster.go]: ok

File: topology/graph/indexer.go [hunk 7]
```diff
@@ -204,7 +204,7 @@ func NewMetadataIndexer(g *Graph, listenerHandler ListenerHandler, m ElementMatc
 		Indexer: NewIndexer(g, listenerHandler, func(n *Node) (kv map[string]interface{}) {
 			if match := n.MatchMetadata(m); match {
 				kv = make(map[string]interface{})
-				if values, err := getFieldsAsArray(n, indexes); err == nil && len(values) > 0 {
+				if values, err := getFieldsAsArray(n, indexes); err == nil && len(indexes) == len(values) {
```
Comment (lebauce) [inline topology/graph/indexer.go]: This doesn't seem related to the compilation of the tests

Comment (hunchback) [reply topology/graph/indexer.go]: removed from the PR

**Comment by lebauce** [comment]
run skydive-functional-tests-backend-elasticsearch

**Comment by lebauce** [comment]
run skydive-ppc64le-tests
